lib/Doctrine/Common/DataFixtures/ReferenceRepository.php

```diff
@@ -151,25 +184,43 @@ public function addReference($name, $object)
      * named by $name
      *
      * @param string $name
+     * @param string|null $class
      *
      * @return object
      *
+     * @template T of object
+     * @psalm-param class-string<T>|null $class
+     * @psalm-return $class is null ? object : T
+     *
      * @throws OutOfBoundsException - if repository does not exist.
      */
-    public function getReference($name)
+    public function getReference($name, $class = null) // NEXT_MAJOR: Make $class mandatory
     {
-        if (! $this->hasReference($name)) {
-            throw new OutOfBoundsException(sprintf('Reference to "%s" does not exist', $name));
+        if (! $this->hasReference($name, $class)) {
+            // For BC, to be removed in next major.
+            if ($class === null) {
+                throw new OutOfBoundsException(sprintf('Reference to "%s" does not exist', $name));
+            }
+
+            throw new OutOfBoundsException(sprintf('Reference to "%s" for class "%s" does not exist', $name, $class));
+        }
+
+        $reference = $class === null
+            ? $this->references[$name] // For BC, to be removed in next major.
+            : $this->referencesByClass[$class][$name];
+
+        if ($class === null) { // For BC, to be removed in next major.
+            $class = get_class($reference);
         }
 
-        $reference = $this->references[$name];
-        $meta      = $this->manager->getClassMetadata(get_class($reference));
+        $meta = $this->manager->getClassMetadata($class);
 
-        if (! $this->manager->contains($reference) && isset($this->identities[$name])) {
-            $reference               = $this->manager->getReference(
-                $meta->name,
-                $this->identities[$name]
-            );
+        $identity = $class === null
```

Comment:
it is never null at that point as you assign it above

Reply:
Fixed by swapping the line to have a consistent logic